source/common/http/filter/squash_filter.cc

```diff
@@ -0,0 +1,340 @@
+  }
+
+  return FilterHeadersStatus::StopIteration;
+}
+
+FilterDataStatus SquashFilter::decodeData(Buffer::Instance&, bool) {
+  if (state_ == State::Initial) {
+    return FilterDataStatus::Continue;
+  } else {
+    return FilterDataStatus::StopIterationAndBuffer;
+  }
+}
+
+FilterTrailersStatus SquashFilter::decodeTrailers(HeaderMap&) {
+  if (state_ == State::Initial) {
+    return FilterTrailersStatus::Continue;
+  } else {
+    return FilterTrailersStatus::StopIteration;
+  }
+}
+
+void SquashFilter::setDecoderFilterCallbacks(StreamDecoderFilterCallbacks& callbacks) {
+  decoder_callbacks_ = &callbacks;
+}
+
+void SquashFilter::onSuccess(MessagePtr&& m) {
+  in_flight_request_ = nullptr;
+
+  switch (state_) {
+
+  case State::Initial: {
+    // Should never happen.
+    ENVOY_LOG(info, "Squash: received send success callback when no request is in progress");
+    break;
+  }
+  case State::CreateConfig: {
+    // get the config object that was created
+    if (Utility::getResponseStatus(m->headers()) != enumToInt(Code::Created)) {
+      ENVOY_LOG(info, "Squash: can't create attachment object. status {} - not squashing",
+                m->headers().Status()->value().c_str());
+      doneSquashing();
+    } else {
+      state_ = State::CheckAttachment;
+
+      std::string debugAttachmentId;
+      try {
+        Json::ObjectSharedPtr json_config = getJsonBody(std::move(m));
+        debugAttachmentId =
+            json_config->getObject("metadata", true)->getString("name", EMPTY_STRING);
+      } catch (Json::Exception&) {
+        debugAttachmentId = EMPTY_STRING;
+      }
+
+      if (debugAttachmentId.empty()) {
+        ENVOY_LOG(info, "Squash: failed to parse debug attachment object - check server settings.");
+        doneSquashing();
+      } else {
+        debugAttachmentPath_ = POST_ATTACHMENT_PATH + debugAttachmentId;
+        pollForAttachment();
+      }
+    }
+
+    break;
+  }
+  case State::CheckAttachment: {
+
+    std::string attachmentstate;
+    try {
+      Json::ObjectSharedPtr json_config = getJsonBody(std::move(m));
+      attachmentstate = json_config->getObject("status", true)->getString("state", EMPTY_STRING);
+    } catch (Json::Exception&) {
+      // no state yet.. leave it empty for the retry logic.
+    }
+
+    bool attached = attachmentstate == ATTACHED_STATE;
+    bool error = attachmentstate == ERROR_STATE;
+    bool finalstate = attached || error;
+
+    if (finalstate) {
+      doneSquashing();
+    } else {
+      retry();
+    }
+    break;
+  }
+  }
+}
+
+void SquashFilter::onFailure(AsyncClient::FailureReason) {
+  // in_flight_request_ will be null if we are called inline of async client send()
+  bool request_created = in_flight_request_ != nullptr;
+  in_flight_request_ = nullptr;
+  switch (state_) {
+  case State::Initial: {
+    // Should never happen.
+    ENVOY_LOG(info, "Squash: received send failure callback when no request is in progress");
+    break;
```

Comment:
totally optional but you could avoid this by having thin shim classes which implement Http::AsyncClient::Callbacks and call an appropriately named function (onConfigSuccess ... onPollFail) in the parent class.  Those functions could just then ASSERT the client is in the correct state for clarity

Reply:
neat, didn't connect the dots here.

Reply:
Does this make sense?
```
class SquashFilter{
...
...
  class CreateConfigCallback : public AsyncClient::Callbacks {
  public:
    CreateConfigCallback(SquashFilter* parent) : parent_(parent){}
    // Http::AsyncClient::Callbacks
    void onSuccess(MessagePtr&&) override {parent->onCreateConfigSuccess()}
    void onFailure(AsyncClient::FailureReason) override {parent->onCreateConfigFailure()}
  private:
    SquashFilter* parent_;
  } configCallback;
...
...
};
```
should I separate the methods out to the cc file?

Reply:
that looks good, minus the configCallback at the end.    I believe Envoy style allows simple functions to be in the .h and I'm fine with it here, as I think it increases clarity.

Reply:
What about this instead:
```
class AsyncClientCallbackShim : public AsyncClient::Callbacks {
public:
  AsyncClientCallbackShim(std::function<void(MessagePtr&&)>&& on_success, std::function<void(AsyncClient::FailureReason)>&& on_fail) 
  : on_success_(on_success), on_fail_(on_fail) {}
  // Http::AsyncClient::Callbacks
  void onSuccess(MessagePtr&& m) override {on_success_(std::forward<MessagePtr>(m));}
  void onFailure(AsyncClient::FailureReason f) override {on_fail_(f);}
private:
  std::function<void(MessagePtr&&)> on_success_;
  std::function<void(AsyncClient::FailureReason)> on_fail_;
};
```
and then the squash filter will look like this:
```
class SquashFilter {
...
  AsyncClientCallbackShim createConfigCallback_;
...
}
SquashFilter::SquashFilter(...) : ...,      createConfigCallback_(std::bind(&SquashFilter::onCreateConfigSuccess, this, _1), std::bind(&SquashFilter::onFailure, this, _1)),

```

Can even go somewhere as a util class?